File: backtracking/subsets_2/main.go
package main

import (
  "fmt"
)

func subsetsWithDup(nums []int) *[][]int {
  res := new([][]int)
  startingSubset := []int{}
  startingIdx := 0

  backtrack(startingSubset, nums, startingIdx, res)

  return res
}

func backtrack(subset, nums []int, idx int, res *[][]int) {
  if idx > len(nums) -1 {
    tmp := make([]int, len(subset))
    copy(tmp, subset)
    *res = append(*res, tmp)
    return
  }

  subset = append(subset, nums[idx])
  backtrack(subset, nums, idx+1, res)
  subset = subset[:len(subset)-1]
  for idx+1 < len(nums) && nums[idx] == nums[idx+1] {
    idx++
  }

  backtrack(subset, nums, idx+1, res)

}


func main() {
  nums := []int{1,2,2}
  res := subsetsWithDup(nums)
  fmt.Println(res)
}

File: backtracking/subsets_2/main.py

import copy

class Solution:
    def subsetsWithDup(self, nums: list[int]) -> list[list[int]]:
        res = []
        nums.sort()
        def backtrack(i: int, subset: list[int]):
            print(subset)
            if i >= len(nums):
                res.append(subset.copy())
                return

            # pick
            subset.append(nums[i])
            backtrack(i+1, subset)

            # skip
            subset.pop()
            j = i
            while j < len(nums) and nums[j] == nums[i]:
                j += 1

            backtrack(j, subset)
        backtrack(0, [])

        return res

sol = Solution()
nums = [1,2,1]

print(sol.subsetsWithDup(nums))
#assert sol.subsetsWithDup(nums), [[],[1],[1,2],[1,1],[1,2,1],[2]]
